extensions/transport_sockets/tls/ssl_socket.cc

```diff
@@ -247,6 +307,16 @@ void SslSocket::shutdownSsl() {
   }
 }
 
+void SslSocket::asyncCb() {
+  ENVOY_CONN_LOG(debug, "SSL async done!", callbacks_->connection());
+
+  // We lose the return value here, so might consider propagating it with an event
+  // in case we run into "Close" result from the handshake handler.
+  if (!handshake_complete_) {
+    doHandshake();
```

Comment:
You should check the return value from doHandshake(), and if it's ` PostIoAction::Close)`, close the SSL connection (call `callbacks_->connection().close(Network::ConnectionCloseType::FlushWrite)`).